### PR TITLE
Allow specifying images for components

### PR DESCRIPTION
**Pull request by kvaps**
Ability to specify component images using environment variables, example:

```yaml
        env:
        - name: OPERATOR_IMAGE
          value: dev-registry.deckhouse.io/sys/deckhouse-oss:292f42af69a61b1252a048c038e1b91834163f7effffd04c6e1f91f2-1660910967128
        - name: VIRT_API_IMAGE
          value: dev-registry.deckhouse.io/sys/deckhouse-oss:1a3eba570cf0d5b9e866f68aadd49b6fbcb95c201d3534f9b00ecbe4-1660911041079
        - name: VIRT_CONTROLLER_IMAGE
          value: dev-registry.deckhouse.io/sys/deckhouse-oss:017823671d1a2a39a94c460670f5b3e330b8937ba3b6925e38c4e051-1660911085211
        - name: VIRT_HANDLER_IMAGE
          value: dev-registry.deckhouse.io/sys/deckhouse-oss:df4ecb86878ddf246bc67f3e89902730ea9ad0ee0dafa760d5956b00-1660910996537
        - name: VIRT_LAUNCHER_IMAGE
          value: dev-registry.deckhouse.io/sys/deckhouse-oss:ae26238adfc5823e96f1dff6ece248b424a57a59b3e558a1a506cc44-1660911013376
        - name: VIRT_EXPORTPROXY_IMAGE
          value: dev-registry.deckhouse.io/sys/deckhouse-oss:017823671d1a2a39a94c460670f5b3e330b8937ba3b6925e38c4e052-1660911085211
        - name: VIRT_EXPORTSERVER_IMAGE
          value: dev-registry.deckhouse.io/sys/deckhouse-oss:017823671d1a2a39a94c460670f5b3e330b8937ga3b6925e38c4e052-1660911085211
        - name: KUBEVIRT_VERSION
          value: v0.55.0
```
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubevirt/kubevirt/issues/8302

**Special notes for your reviewer**:

**Release note**:

```release-note
Allow specifying images for components
```

### CODE REVIEW

File: pkg/virt-operator/resource/generate/components/daemonsets.go [hunk 1]
```diff
@@ -22,10 +22,12 @@ const (
 	kubeletPodsPath = "/var/lib/kubelet/pods"
 )
 
-func NewHandlerDaemonSet(namespace string, repository string, imagePrefix string, version string, launcherVersion string, productName string, productVersion string, productComponent string, pullPolicy corev1.PullPolicy, migrationNetwork *string, verbosity string, extraEnv map[string]string) (*appsv1.DaemonSet, error) {
+func NewHandlerDaemonSet(namespace string, imageName string, repository string, imagePrefix string, version string, launcherImage string, launcherVersion string, productName string, productVersion string, productComponent string, pullPolicy corev1.PullPolicy, migrationNetwork *string, verbosity string, extraEnv map[string]string) (*appsv1.DaemonSet, error) {
 
 	deploymentName := VirtHandlerName
-	imageName := fmt.Sprintf("%s%s", imagePrefix, deploymentName)
+	if imageName == "" {
```
Comment (enp0s3) [inline pkg/virt-operator/resource/generate/components/daemonsets.go:28]: We prefer not to re-use a variable that was passed as an argument, can we do something like this instead:
```
imageName := VirtHandlerImageName
if len(imageName) == 0 {
  imageName = fmt.Sprintf("%s%s", imagePrefix, deploymentName) 
}
```

File: pkg/virt-operator/resource/generate/components/daemonsets.go [hunk 2]
```diff
@@ -83,13 +85,16 @@ func NewHandlerDaemonSet(namespace string, repository string, imagePrefix string
 	// nodelabeller currently only support x86
 	if virtconfig.IsAMD64(runtime.GOARCH) {
 		launcherVersion = AddVersionSeparatorPrefix(launcherVersion)
+		if launcherImage == "" {
```
Comment (enp0s3) [inline pkg/virt-operator/resource/generate/components/daemonsets.go:88]: Same here

File: pkg/virt-operator/resource/generate/components/deployments.go [hunk 3]
```diff
@@ -302,10 +302,12 @@ func newPodAntiAffinity(key, topologyKey string, operator metav1.LabelSelectorOp
 	}
 }
 
-func NewApiServerDeployment(namespace string, repository string, imagePrefix string, version string, productName string, productVersion string, productComponent string, pullPolicy corev1.PullPolicy, verbosity string, extraEnv map[string]string) (*appsv1.Deployment, error) {
+func NewApiServerDeployment(namespace string, imageName string, repository string, imagePrefix string, version string, productName string, productVersion string, productComponent string, pullPolicy corev1.PullPolicy, verbosity string, extraEnv map[string]string) (*appsv1.Deployment, error) {
 	podAntiAffinity := newPodAntiAffinity(kubevirtLabelKey, kubernetesHostnameTopologyKey, metav1.LabelSelectorOpIn, []string{VirtAPIName})
 	deploymentName := VirtAPIName
-	imageName := fmt.Sprintf("%s%s", imagePrefix, deploymentName)
+	if imageName == "" {
```
Comment (enp0s3) [inline pkg/virt-operator/resource/generate/components/deployments.go:308]: Same here

File: pkg/virt-operator/resource/generate/components/deployments.go [hunk 4]
```diff
@@ -372,10 +374,20 @@ func NewApiServerDeployment(namespace string, repository string, imagePrefix str
 	return deployment, nil
 }
 
-func NewControllerDeployment(namespace string, repository string, imagePrefix string, controllerVersion string, launcherVersion string, exportServerVersion string, productName string, productVersion string, productComponent string, pullPolicy corev1.PullPolicy, verbosity string, extraEnv map[string]string) (*appsv1.Deployment, error) {
+func NewControllerDeployment(namespace string, imageName string, repository string, imagePrefix string, controllerVersion string, launcherImage string, launcherVersion string, exporterImage string, exportServerVersion string, productName string, productVersion string, productComponent string, pullPolicy corev1.PullPolicy, verbosity string, extraEnv map[string]string) (*appsv1.Deployment, error) {
 	podAntiAffinity := newPodAntiAffinity(kubevirtLabelKey, kubernetesHostnameTopologyKey, metav1.LabelSelectorOpIn, []string{VirtControllerName})
 	deploymentName := VirtControllerName
-	imageName := fmt.Sprintf("%s%s", imagePrefix, deploymentName)
+	if imageName == "" {
```
Comment (enp0s3) [inline pkg/virt-operator/resource/generate/components/deployments.go:380]: Same here

**Review by enp0s3 — COMMENTED** [review]
@kvaps Hi, thank you for the PR! 

I have a small request -
We prefer not to re-use a variable that was passed as an argument, can we do something like this instead:
```
imageName := VirtHandlerImageName
if len(imageName) == 0 {
  imageName = fmt.Sprintf("%s%s", imagePrefix, deploymentName) 
}
```
Since it is being used in multiple places in the PR, can you please address them?

**Comment by kvaps** [comment]
Sure, makes sense, thanks for suggestion

**Comment by iholder-redhat** [comment]
Thanks for your efforts @kvaps! Looks great!

Your PR is currently marked as a draft PR, so I assume you're still working on it. When it's done please mark it a "ready to review". You're also welcome to ping me to make sure I review it asap.

Some thoughts / questions:
1. The example from your PR description - where is it being defined? In which manifest?
2. It's definitely not your fault, but with the way env variables are currently used they're impossible to mock. Therefore, the unit test you've introduced actually defines real env variables for whoever is running them. Ideally, we'd use some kind of interface that would consume env variables on production and fake variables on unit tests. I think it shouldn't block your PR as it's outside of scope, I just want to point that out here so it we keep track of it. I'll gladly take care of that in a follow-up PR.

**Comment by iholder-redhat** [comment]
We can also run one relevant lane to see where we stand.
/test pull-kubevirt-e2e-k8s-1.23-operator

**Comment by iholder-redhat** [comment]
@kvaps any reason it's being pushed to `v0.57` and not to main branch?

**Comment by kvaps** [comment]
> @kvaps any reason it's being pushed to v0.57 and not to main branch?

For [some of my PRs](https://github.com/kubevirt/kubevirt/pulls/kvaps) which are not going to be merged into main right now I prefer to base on release branch, because it's easier to generate patch from them and add them into [our platform](https://github.com/deckhouse/deckhouse/tree/c4a3911a0d722d1d35aee09ef175f2ca1666d006/modules/900-kubevirt/images/artifact/patches).

> Your PR is currently marked as a draft PR, so I assume you're still working on it. When it's done please mark it a "ready to review". You're also welcome to ping me to make sure I review it asap.

Yeah, currently this is just a simple patch which is breaking the existing functionality. Which is not good.
I'm going to rework it on in the next week.

>  The example from your PR description - where is it being defined? In which manifest?

This is virt-operator manifest, we have similar logic for specifying [image digests in kubevirt](https://github.com/kubevirt/kubevirt/blob/0484054275ffebb33142204e82a056730b53ca5e/pkg/virt-operator/util/config.go#L254-L265) and [image names in cdi-operator](https://github.com/kubevirt/containerized-data-importer/blob/73ac262ec4d4da79881ec1e129400810167a4d0e/pkg/operator/resources/operator/operator.go#L326-L371)

>    It's definitely not your fault, but with the way env variables are currently used they're impossible to mock. Therefore, the unit test you've introduced actually defines real env variables for whoever is running them. Ideally, we'd use some kind of interface that would consume env variables on production and fake variables on unit tests. I think it shouldn't block your PR as it's outside of scope, I just want to point that out here so it we keep track of it. I'll gladly take care of that in a follow-up PR.

Thanks I'll take this into account

**Comment by iholder-redhat** [comment]
@kvaps thanks for your quick reply!
Please let me know if I can help you push this forward in any way.

> > @kvaps any reason it's being pushed to v0.57 and not to main branch?
> 
> For [some of my PRs](https://github.com/kubevirt/kubevirt/pulls/kvaps) which are not going to be merged into main right now I prefer to base on release branch, because it's easier to generate patch from them and add them into [our platform](https://github.com/deckhouse/deckhouse/tree/c4a3911a0d722d1d35aee09ef175f2ca1666d006/modules/900-kubevirt/images/artifact/patches).

I think the right thing to do is to first merge this into main branch, then backport it into `v0.57`. Since it's the latest release, there's a very big probability that an automatic backport would be enough.

The problem with pushing changes straight to release branches is that they won't automatically continue to the next version (`v0.58` and so on). I suggest to open a new PR to push it into main branch, then I'll help you backport the PR to v0.57 once it's merged. WDYT?

**Comment by stu-gott** [comment]
Hi @kvaps I see this PR is still a draft. Do you need help getting this into shape in some way?

**Comment by kvaps** [comment]
Hi @stu-gott. I'm sorry, I have no enough time for this PR right now, but it is still in a row.
It was just moved from phase 1, to phase 2 in our internal roadmap of implementing KubeVirt into Deckhouse platform
- https://github.com/deckhouse/deckhouse/issues/602

**Comment by stu-gott** [comment]
Thanks for clarifying @kvaps. This PR also aligns with our priorities. Would it be OK if we contributed changes to this PR?

**Comment by kvaps** [comment]
Sure, you can base on 118c5a2a8cde7bd044430c5d97b585211f36ff01, the current version simpler but incompatible

**Comment by iholder101** [comment]
Hey @kvaps!
As https://github.com/kubevirt/kubevirt/pull/8673 is already merged, do you think this PR can be closed?

Thanks

**Comment by kvaps** [comment]
Yeah I think so

/close

**Pull request closed without merging**